db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -3957,13 +3957,21 @@ void DBImpl::GetSnapshotContext(
   *snapshot_seqs = snapshots_.GetAll(earliest_write_conflict_snapshot);
 }
 
-Status DBImpl::WaitForCompact(bool wait_unscheduled) {
-  // Wait until the compaction completes
+Status DBImpl::WaitForCompact(bool abort_on_pause) {
   InstrumentedMutexLock l(&mutex_);
+  if (shutting_down_.load(std::memory_order_acquire)) {
+    return Status::ShutdownInProgress();
+  }
+  if (bg_work_paused_ && abort_on_pause) {
```

Comment:
"these" = this case also. The idea of abort on pause is to return Aborted if someone has called PauseBackgroundWork(). Whether that's at the beginning of WaitForCompact or after it's already waiting, it should still happen (in case it's needed to avoid a deadlock).

Reply:
You should be able to minimize copy-paste by rewriting the loop to `for(;;) { ... }` with a series of early return conditions inside. (Could also fix the double-read, possible TOCTOU bug, on `error_handler_.GetBGError()`. It might not be a bug at the moment due to holding DB mutex, but it's precarious nonetheless.)

Reply:
Maybe I should learn more about how really mutex works. I thought if `PauseBackgroundWork()` is called while waiting or any point in WaitForCompact(), it would wait to get the lock before pausing, so by the time it actually starts pausing, `WaitForCompact()` would have been done with waiting (or returned by other status) because lock must have been released.